library/src/tests/diagnostics.rs

```diff
@@ -40,3 +40,119 @@ fn check_operations_are_equal() {
+            import Microsoft.Quantum.Diagnostics.StopCountingOperation;
+
+            operation op1() : Unit {}
+            operation op2() : Unit { op1(); }
+            StartCountingOperation(op1);
+            StartCountingOperation(op2);
+            (StopCountingOperation(op1), StopCountingOperation(op2))
+        }",
+        &Value::Tuple([Value::Int(0), Value::Int(0)].into()),
+    );
+}
+
+#[test]
+fn check_stop_counting_operation_without_start() {
+    test_expression(
+        "{
+            import Microsoft.Quantum.Diagnostics.StopCountingOperation;
+
+            operation op1() : Unit {}
+            StopCountingOperation(op1)
+        }",
+        &Value::Int(-1),
+    );
+}
+
+#[test]
+fn check_counting_operation_differentiates_between_body_adj_ctl() {
+    test_expression(
+        "{
+            import Microsoft.Quantum.Diagnostics.StartCountingOperation;
+            import Microsoft.Quantum.Diagnostics.StopCountingOperation;
+
+            operation op1() : Unit is Adj + Ctl {}
+            StartCountingOperation(op1);
+            StartCountingOperation(Adjoint op1);
+            StartCountingOperation(Controlled op1);
+            op1();
+            Adjoint op1(); Adjoint op1();
+            Controlled op1([], ()); Controlled op1([], ()); Controlled op1([], ());
+            (StopCountingOperation(op1), StopCountingOperation(Adjoint op1), StopCountingOperation(Controlled op1))
+        }",
+        &Value::Tuple([Value::Int(1), Value::Int(2), Value::Int(3)].into()),
+    );
+}
+
+#[test]
+fn check_start_stop_counting_function_called_3_times() {
+    test_expression(
+        "{
+            import Microsoft.Quantum.Diagnostics.StartCountingFunction;
+            import Microsoft.Quantum.Diagnostics.StopCountingFunction;
+
+            function f1() : Unit {}
+            function f2() : Unit { f1(); }
+            StartCountingFunction(f1);
+            StartCountingFunction(f2);
+            f1(); f1(); f2();
+            (StopCountingFunction(f1), StopCountingFunction(f2))
+        }",
+        &Value::Tuple([Value::Int(3), Value::Int(1)].into()),
+    );
+}
+
+#[test]
+fn check_start_stop_counting_function_called_0_times() {
+    test_expression(
+        "{
+            import Microsoft.Quantum.Diagnostics.StartCountingFunction;
+            import Microsoft.Quantum.Diagnostics.StopCountingFunction;
+
+            function f1() : Unit {}
+            function f2() : Unit { f1(); }
+            StartCountingFunction(f1);
+            StartCountingFunction(f2);
+            (StopCountingFunction(f1), StopCountingFunction(f2))
+        }",
+        &Value::Tuple([Value::Int(0), Value::Int(0)].into()),
+    );
+}
+
+#[test]
+fn check_stop_counting_function_without_start() {
+    test_expression(
+        "{
+            import Microsoft.Quantum.Diagnostics.StopCountingFunction;
+
+            function f1() : Unit {}
+            StopCountingFunction(f1)
+        }",
+        &Value::Int(-1),
+    );
+}
```

Comment:
Should there be some tests that cover the built-in intrinsics as well, like `X`, etc.?

Reply:
I don't think it should be needed, from the perspective of the evaluator every callable is just a resolution (item id plus package id). The evaluator infra makes sure all callables are treated equal.